storage/blockchain/src/tables.rs

```diff
@@ -14,335 +14,38 @@
-    Outputs => 7,
-    PrunableTxBlobs => 8,
-    RctOutputs => 9,
-    TxBlobs => 10,
-    TxIds => 11,
-    TxHeights => 12,
-    TxOutputs => 13,
-    TxUnlockTime => 14,
-}
-
-//---------------------------------------------------------------------------------------------------- Table macro
-/// Create all tables, should be used _once_.
-///
-/// Generating this macro once and using `$()*` is probably
-/// faster for compile times than calling the macro _per_ table.
-///
-/// All tables are zero-sized table structs, and implement the `Table` trait.
-///
-/// Table structs are automatically `CamelCase`,
-/// and their static string names are automatically `snake_case`.
-macro_rules! tables {
-    (
-        $(
-            $(#[$attr:meta])* // Documentation and any `derive`'s.
-            $table:ident,     // The table name + doubles as the table struct name.
-            $key:ty =>        // Key type.
-            $value:ty         // Value type.
-        ),* $(,)?
-    ) => {
-        paste::paste! { $(
-            // Table struct.
-            $(#[$attr])*
-            // The below test show the `snake_case` table name in cargo docs.
-            #[doc = concat!("- Key: [`", stringify!($key), "`]")]
-            #[doc = concat!("- Value: [`", stringify!($value), "`]")]
-            ///
-            /// ## Table Name
-            /// ```rust
-            /// # use cuprate_blockchain::{*,tables::*};
-            /// use cuprate_database::Table;
-            #[doc = concat!(
-                "assert_eq!(",
-                stringify!([<$table:camel>]),
-                "::NAME, \"",
-                stringify!([<$table:snake>]),
-                "\");",
-            )]
-            /// ```
-            #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-            #[derive(Copy, Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
-            pub struct [<$table:camel>];
-
-            // Implement the `Sealed` in this file.
-            // Required by `Table`.
-            impl private::Sealed for [<$table:camel>] {}
-
-            // Table trait impl.
-            impl Table for [<$table:camel>] {
-                const NAME: &'static str = stringify!([<$table:snake>]);
-                type Key = $key;
-                type Value = $value;
-            }
-        )* }
-    };
-}
-
 //---------------------------------------------------------------------------------------------------- Tables
 // Notes:
 // - Keep this sorted A-Z (by table name)
 // - Tables are defined in plural to avoid name conflicts with types
 // - If adding/changing a table also edit:
 //   - the tests in `src/backend/tests.rs`
 //   - `call_fn_on_all_tables_or_early_return!()` macro in `src/open_tables.rs`
-tables! {
+cuprate_database::define_tables! {
     /// Serialized block blobs (bytes).
     ///
     /// Contains the serialized version of all blocks.
-    BlockBlobs,
+    0 => BlockBlobs,
     BlockHeight => BlockBlob,
```

Comment:
This will be easy to do once this is stabilized: https://veykril.github.io/tlborm/decl-macros/minutiae/metavar-expr.html#indexdepth for now this is fine